core-rust/state-manager/src/accumulator_tree/tree_builder.rs

```diff
@@ -0,0 +1,247 @@
+    current_len: usize,
+    phantom_data: PhantomData<N>,
+}
+
+impl<'s, S: AccuTreeStore<M>, M: Merklizable> AccuTree<'s, S, M> {
+    /// Creates a new accumulator tree backed by the given store, which is assumed to currently hold
+    /// the given number of leaves.
+    pub fn new(store: &'s mut S, current_len: usize) -> Self {
+        Self {
+            store,
+            current_len,
+            phantom_data: PhantomData,
+        }
+    }
+
+    /// Appends the given batch of new leaves to the tree.
+    pub fn append(&mut self, leaves: Vec<M>) {
+        let new_leaf_count = leaves.len();
+        if new_leaf_count == 0 {
+            return;
+        }
+
+        let mut previous_slice_levels = if self.current_len == 0 {
+            vec![].into_iter()
+        } else {
+            let previous_slice = self
+                .store
+                .get_tree_slice(self.current_len)
+                .expect("no slice ending at given version found");
+            previous_slice.levels.into_iter()
+        };
+
+        let target_length = self.current_len + new_leaf_count;
+
+        let mut from = self.current_len;
+        let mut higher_level_nodes = leaves;
+
+        let mut levels = Vec::new();
+        let target_height = usize::BITS - (target_length - 1).leading_zeros();
+        for _ in 0..target_height {
+            let previous_slice_level = previous_slice_levels.next();
+            let left_sibling_cache = if from % 2 == 0 {
+                None
+            } else {
+                let previous_slice_level = previous_slice_level.unwrap();
+                Some(
+                    RelativeNodeAccess::for_successor_offset(previous_slice_level, from)
+                        .into_node(from - 1),
+                )
+            };
+            let lower_level_access = RelativeNodeAccess::for_offset(
+                from,
+                TreeSliceLevel::new(left_sibling_cache, higher_level_nodes),
+            );
+
+            let to = (from + lower_level_access.slice_level.nodes.len() + 1) / 2;
+            from /= 2;
+            higher_level_nodes = (from..to)
+                .map(|level_index| level_index * 2)
+                .map(|lower_level_index| {
+                    M::accumulate(
+                        lower_level_access.get(lower_level_index),
+                        lower_level_access.get(lower_level_index + 1),
+                    )
+                })
+                .collect();
+
+            levels.push(lower_level_access.into_level());
+        }
+        levels.push(TreeSliceLevel::new(None, higher_level_nodes));
+
+        self.current_len = target_length;
+        self.store
+            .put_tree_slice(self.current_len, TreeSlice::new(levels));
+    }
+}
+
+enum RelativeIndex {
+    CachedLeftSibling,
+    ActualStoredNode(usize),
+    ZeroRightSibling,
+}
+
+struct RelativeNodeAccess<M: Merklizable> {
+    offset: usize,
+    slice_level: TreeSliceLevel<M>,
+    zero: M,
```

Comment:
I have this `get()` function, which is supposed to return a reference :/ (the caller only wants reference too)
If I make it return a value (at the cost of `clone()` here and there), then I can totally get rid of `zero` (not needing `PhantomData` either). Is it worth it?

Reply:
Oops, sorry I missed that `get`. Nope, definitely not worth it.

Reply:
> Nope, definitely not worth it.

Hah, you are the deadliest enemy of `.clone()` :D
I get it and I respect it.
But also: at a price of (in practice) 32-byte memcpy, we could get rid of `zero: M` (to which many future readers will react "wait, why the guy has to store a zero here?!!".
And we don't even know if that `.clone()` costs anything in runtime (maybe the compiler notices "hey let's reuse this mem").
And we also kinda disregard that `zero: M` seen in this line requires a... 32-byte mem-cpy when we create an instance of `RelativeNodeAccess` :) (so we save a potential >=0 memcpy there at the price of guaranteed 1 memcpy here)

So to me, maybe it's worth it to return a val from `get()` after all?

Reply:
> "wait, why the guy has to store a zero here?!!"
This indeed is confusing first time encountering it.

Probably this might be towards the right solution: https://docs.rs/generic_static/latest/generic_static/ ?

I'm more worried about the `slice_level: TreeSliceLevel<M>` than `zero` when creating a new instance. I imagine an "iterator"-like structure would not own a slice of data, but be used alongside the data structure itself.

Reply:
I remember when I first saw `lazy_static`. I thought "wow, there are more people than just me who try to "fix" rust!". And surely there are :) Let me check that `generic_static` thing for `M::zero()` purposes. _(of course it does not come for free; I see there is a rwLock involved + a hashmap access - which makes me think it is mostly meant for caching expensive objects, not 32-bytes [of "fill 0"])_

> I imagine an "iterator"-like structure

I am genuinely curious to learn that pattern.

But my point was: should we be actively worried about [logical] "moves" of data when we create helper objects in rust? How do even know that compiler does not simply place things at the right mem offsets, so as to make the `TreeSliceLevel::new(...)` 0-cost?
_(I remember worrying about allocation [or rather GC] in certain java OO-heavy logic, until I saw that TLABs optimize all of it away - as if java had on-heap structs)_

I can feel your deep expertise on runtime performance. I totally see how many aspects of this PR could be micro-optimized.
But we cannot focus on everything at once. I think version 1 needs to focus on readability - so your comment on `zero: M,` being "weird" is very valid and I want to improve it. It is of course important to also avoid anything obviously sub-optimal (like doing O(n^2) when not necessary; or rewriting growing data volumes in storage). But optimizing anything more than that can be postponed - ideally after we identify bottlenecks.

Reply:
There is no clear "best" way here, so let me summarize rough options and gather more opinions:

The problem:
A "logic helper" struct `RelativeNodeAccess<M>` has a `get M by index` method, which sometimes legitimately wants to return a meaningful "zero" object. Zero can be constructed statically (for a known `M` type). The callers of that `get` only want a reference.

How to do it?

A) Construct `M::zero()` on creation of `RelativeNodeAccess<M>` and store it as a [weird] part of that struct, in order to be able to return `&M` from `get()`. _(this option is currently seen in this PR)_

B) Return `M::zero()` directly from `get()` [even though no caller wants a value, they only want a ref]

C) Put `M::zero()` into a "generic static" structure [i.e. introduce dep https://docs.rs/generic_static/latest/generic_static/ and accept its runtime RwLock + HashMap implications], then easily return `&M` from `get()`.

@LukasGasior1 wdyt?

Reply:
I'd be leaning towards B), but happy to keep it as is too. Introducing generic_static seems a bit of an overkill here

Reply:
Alright, thanks 👍 

Seems like A) is the good compromise option that everyone can live with.
I have added a comment on `zero: M` field in struct, so that it does not feel that weird in future.

Reply:
> or rewriting growing data volumes in storage). But optimizing anything more than that can be postponed - ideally after we identify bottlenecks.
I agree, not worth optimizing the constant unless something outrageous (which is not the case here). My initial comment was missing the `get` and assumed that `zero: M` was needed just for the type transport (`PhantomData`) but I was wrong :).

I tend towards A) as well.